src/ecmascript/spidermonkey/console.c

```diff
@@ -0,0 +1,150 @@
+#include "session/task.h"
+#include "terminal/tab.h"
+#include "terminal/terminal.h"
+#include "util/conv.h"
+#include "util/memory.h"
+#include "util/string.h"
+#include "viewer/text/draw.h"
+#include "viewer/text/form.h"
+#include "viewer/text/link.h"
+#include "viewer/text/vs.h"
+
+#include <time.h>
+#include "document/renderer.h"
+#include "document/refresh.h"
+#include "terminal/screen.h"
+
+#define DEBUG 0
+
+static bool console_get_property(JSContext *ctx, JS::HandleObject hobj, JS::HandleId hid, JS::MutableHandleValue hvp);
+
+JSClassOps console_ops = {
+	JS_PropertyStub, nullptr,
+	console_get_property, JS_StrictPropertyStub,
+	nullptr, nullptr, nullptr
+};
+
+/* Each @console_class object must have a @window_class parent.  */
+const JSClass console_class = {
+	"console",
+	JSCLASS_HAS_PRIVATE,
+	&console_ops
+};
+
+/* Tinyids of properties.  Use negative values to distinguish these
+ * from array indexes (even though this object has no array elements).
+ * ECMAScript code should not use these directly as in document[-1];
+ * future versions of ELinks may change the numbers.  */
+enum console_prop {
+	JSP_LOCALSTORAGE_LOC   = -1,
+	JSP_LOCALSTORAGE_REF   = -2,
+	JSP_LOCALSTORAGE_TITLE = -3,
+	JSP_LOCALSTORAGE_URL   = -4,
+};
+/* "cookie" is special; it isn't a regular property but we channel it to the
+ * cookie-module. XXX: Would it work if "cookie" was defined in this array? */
+const JSPropertySpec console_props[] = {
+	{ "location",	JSP_LOCALSTORAGE_LOC,	JSPROP_ENUMERATE },
+	{ "referrer",	JSP_LOCALSTORAGE_REF,	JSPROP_ENUMERATE | JSPROP_READONLY },
+	{ "title",	JSP_LOCALSTORAGE_TITLE,	JSPROP_ENUMERATE }, /* TODO: Charset? */
+	{ "url",	JSP_LOCALSTORAGE_URL,	JSPROP_ENUMERATE },
+	{ NULL }
+};
+
+/* @console_class.getProperty */
+static bool
+console_get_property(JSContext *ctx, JS::HandleObject hobj, JS::HandleId hid, JS::MutableHandleValue hvp)
+{
+	JSObject *parent_win;	/* instance of @window_class */
+	struct view_state *vs;
+	struct document_view *doc_view;
+	struct document *document;
+	struct session *ses;
+
+	///* This can be called if @obj if not itself an instance of the
+	// * appropriate class but has one in its prototype chain.  Fail
+	// * such calls.  */
+	assert(JS_InstanceOf(ctx, parent_win, &window_class, NULL));
+	if_assert_failed return false;
+
+	return true;
+}
+
+static bool console_log(JSContext *ctx, unsigned int argc, JS::Value *vp);
+
+const spidermonkeyFunctionSpec console_funcs[] = {
+	{ "log",		console_log,	 	2 },
+	{ NULL }
+};
+
+static bool
+console_log(JSContext *ctx, unsigned int argc, JS::Value *vp)
+{
+	struct ecmascript_interpreter *interpreter = JS_GetContextPrivate(ctx);
+	struct document_view *doc_view = interpreter->vs->doc_view;
+	struct session *ses = doc_view->session;
+	struct terminal *term = ses->tab->term;
+	struct string *ret = interpreter->ret;
+	struct document *document;
+	JS::CallArgs args = CallArgsFromVp(argc, vp);
+	document = doc_view->document;
+
+	unsigned char log[16384];
```

Comment:
What is log here?

Reply:
Sorry, unused variable. How do I edit Your comments? Can You deleted it right from here?

Reply:
Used is log_fname.